add process and thread addresses

yelhamer opened the pull request:
This PR adds addresses for processes and threads, which is a prerequisite for adding dynamic feature extraction to main.

TaskList:
- [x] add ProcessAddress
- [x] add ThreadAddress
- [x] update the Cape extractor
- [x] update the NullExtractors
- [x] update Feature freeze
- [x] update tests
- [x] fix mypy issues
- [x] decide and implement the file header #1614 
- [ ] add tests for `test_dynamic_freeze.py`

### Checklist

<!-- CHANGELOG.md has a `master (unreleased)` section. Please add bug fixes, new features, breaking changes and anything else you think is worthwhile mentioning in the release notes to this file. -->
- [ ] No CHANGELOG update needed
<!-- Tests prove that your fix/work as expected and ensure it doesn't break on the feature. -->
- [ ] No new tests needed
<!-- Please help us keeping capa documentation up-to-date -->
- [ ] No documentation update needed

williballenthin commented on capa/features/address.py:
```diff
@@ -36,6 +36,58 @@ def __hash__(self):
+
+    def __eq__(self, other):
+        assert isinstance(other, ProcessAddress)
+        if self.ppid > 0:
```
im afraid this check here might lead to some surprising behavior, if one has ppid set and the other does not. are you sure this cannot be simplified to `return (self.pid, self.ppid) == (other.pid, other.ppid)`?

williballenthin commented on capa/features/address.py:
```diff
@@ -36,6 +36,58 @@ def __hash__(self):
+
+    def __init__(self, process: ProcessAddress, tid: int):
+        assert tid >= 0
+        self.ppid = process.ppid
+        self.pid = process.pid
```
```suggestion
        self.process = process
```

williballenthin replied:
unless its important that these fields are broken out?

williballenthin commented on capa/features/extractors/cape/file.py:
```diff
@@ -24,8 +24,10 @@ def get_processes(static: Dict) -> Iterator[ProcessHandle]:
-        yield ProcessHandle(pid=process["pid"], inner=inner)
+        address: ProcessAddress = ProcessAddress(pid=process["pid"], ppid=process["parent_id"])
+        inner: Dict[str, str] = {"name": process["name"]}
+        print(address)
```
```suggestion
```

williballenthin commented on capa/features/extractors/null.py:
```diff
@@ -70,3 +78,49 @@ def get_instructions(self, f, bb):
+
+
+@dataclass
+class NullDynamicFeatureExtractor(DynamicFeatureExtractor):
+    base_address: Address
+    global_features: List[Feature]
+    file_features: List[Tuple[Address, Feature]]
+    processes: Dict[Address, ProcessFeatures]
+
+    def extract_global_features(self):
+        for feature in self.global_features:
+            yield feature, NO_ADDRESS
+
+    def extract_file_features(self):
+        for address, feature in self.file_features:
+            yield feature, address
+
+    def get_processes(self):
+        for address in sorted(self.processes.keys()):
+            yield ProcessHandle(address=address, inner={}, pid=address.pid)
+
+    def extract_process_features(self, p):
+        for addr, feature in self.processes[p.address].features:
+            yield feature, addr
+
+    def get_threads(self, p):
+        for address in sorted(self.processes[p].threads.keys()):
+            yield ThreadHandle(address=address, inner={}, tid=address.pid)
+
+    def extract_thread_features(self, p, t):
+        for addr, feature in self.processes[p.address].threads[t.address].features:
+            yield feature, addr
```
👍🏼

williballenthin commented on capa/features/freeze/__init__.py:
```diff
@@ -135,6 +154,36 @@ class FileFeature(HashableModel):
+        process: the address of the process to which this feature belongs.
+        address: the address at which this feature is found.
+
+    process != address because, e.g., the feature may be found *within* the scope (process).
+    versus right at its starting address.
```
i don't understand this - doesn't the `address` contain the information needed to pick the right process? can you maybe include an example that shows how these are not the same?

yelhamer replied:
isn't `address` the address returned by a feature? if so, the the environment strings extracted return a `NO_ADDRESS` for the moment (and so should the future registry and filename features I believe).
If I'm correct, then should we make the address of those features the same as the process address?

williballenthin replied:
yes, each feature is associated with an address.

when possible, we should associate features with the most specific possible address, like an API call feature associated with a DynamicAnalysisEventAddress (or similar). we should use ProcessAddress whenever the feature can only be related to a process but we're not sure at what specific point. an environment string is a good example of this - the string is related to some process, not a specific event. 

we should use `NO_ADDRESS` when we really have no relation to any specific object. im not sure if there are any good examples of this yet for dynamic analysis. perhaps CAPE version (but this isn't a feature...).

williballenthin replied:
> versus right at its starting address.

to clarify, i dont think we should consider ProcessAddress to refer to a specific point in time, like the event that spawns a process or the moment it springs into existence; rather, it should represent the entire Process over time.

williballenthin commented on capa/features/freeze/__init__.py:
```diff
@@ -364,24 +515,68 @@ def loads(s: str) -> StaticFeatureExtractor:
+    )
+
+
+MAGIC = "capa000".encode("ascii")
```
this is ok for now, but we may want to revert this before merge once we've discussed how to extend the file format appropriately.

williballenthin reviewed (approved):
nice work!

please respond to the inline comments and let me know when we should merge.

yelhamer commented:
I am not sure how to run the git actions tests without merging. the conflicts seem to resolve automatically however:

![image](https://github.com/mandiant/capa/assets/16624109/1c320fb6-ae6d-4cbb-8ebf-09cfa2be5807)

Regardless, the tests succeed on my machine:

![image](https://github.com/mandiant/capa/assets/16624109/490ee9bb-adfa-4f85-8af0-d3209b66f76f)

williballenthin commented:
@yelhamer please address merge conflicts then lets merge

yelhamer commented:
@williballenthin I addressed the conflicts and updated the tests, it should be ready for merge now. If there are any other remarks please let me know and I'll work on them.

williballenthin reviewed (approved):
lgtm
